eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -379,6 +363,51 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
     self ! Peer.OutgoingMessage(msg, peerConnection)
   }
 
+  def acceptOrRejectChannel(msg: Either[protocol.OpenChannel, protocol.OpenDualFundedChannel], d: ConnectedData): State = {
+    val temporaryChannelId = msg.fold(_.temporaryChannelId, _.temporaryChannelId)
+    val msgName = msg.fold(_ => "open_channel", _ => "open_channel2")
+    d.channels.get(TemporaryChannelId(temporaryChannelId)) match {
+      case None =>
+        val channelConfig = ChannelConfig.standard
+        val channelType_opt = msg.fold(_.channelType_opt, _.channelType_opt)
+        val chosenChannelType: Either[ChannelException, SupportedChannelType] = channelType_opt match {
+          // remote explicitly specifies a channel type: we check whether we want to allow it
+          case Some(remoteChannelType) => ChannelTypes.areCompatible(d.localFeatures, remoteChannelType) match {
+            case Some(acceptedChannelType) => Right(acceptedChannelType)
+            case None => Left(InvalidChannelType(temporaryChannelId, ChannelTypes.defaultFromFeatures(d.localFeatures, d.remoteFeatures), remoteChannelType))
+          }
+          // Bolt 2: if `option_channel_type` is negotiated: MUST set `channel_type`
+          case None if Features.canUseFeature(d.localFeatures, d.remoteFeatures, Features.ChannelType) => Left(MissingChannelType(temporaryChannelId))
+          // remote doesn't specify a channel type: we use spec-defined defaults
+          case None => Right(ChannelTypes.defaultFromFeatures(d.localFeatures, d.remoteFeatures))
+        }
+        chosenChannelType match {
+          case Right(channelType) =>
+            val fundingAmount = msg.fold(_.fundingSatoshis, _.fundingAmount)
+            val (channel, localParams) = createNewChannel(nodeParams, d.localFeatures, channelType, isInitiator = false, fundingAmount, None)
+            log.info(s"accepting a new channel with type=$channelType temporaryChannelId=$temporaryChannelId localParams=$localParams")
+            msg match {
+              case Left(openSingleFunder) =>
+                channel ! INPUT_INIT_CHANNEL_NON_INITIATOR(temporaryChannelId, None, dualFunded = false, localParams, d.peerConnection, d.remoteInit, channelConfig, channelType)
+                channel ! openSingleFunder
+              case Right(openDualFunded) =>
+                // NB: we don't add a contribution to the funding amount.
+                channel ! INPUT_INIT_CHANNEL_NON_INITIATOR(temporaryChannelId, None, dualFunded = true, localParams, d.peerConnection, d.remoteInit, channelConfig, channelType)
+                channel ! openDualFunded
+            }
```

Comment:
This `match/case` and the number of `fold` in the method make me think `acceptOrRejectChannel` is not the right factorization.

Maybe just factor the channel type negotiation?

Reply:
Good idea, I'll try that.

Reply:
Done in https://github.com/ACINQ/eclair/pull/2247/commits/3a0d2356c42ca22bee145deacb3c277848be0020, the remaining small duplication between the two handlers is completely acceptable imo.